redis/tests/test_cache.rs

```diff
@@ -0,0 +1,307 @@
+        }
+        let val: Option<String> = redis::cmd("GET")
+            .arg("key_1")
+            .query_async(&mut con)
+            .await
+            .unwrap();
+        assert_eq!(val, None);
+        {
+            let s = con.get_cache_statistics();
+            assert_eq!(s.hit, 1);
+            assert_eq!(s.miss, 1);
+        }
+        let _: () = redis::cmd("SET")
+            .arg("key_1")
+            .arg("1")
+            .query_async(&mut con)
+            .await
+            .unwrap();
+        {
+            let s = con.get_cache_statistics();
+            assert_eq!(s.hit, 1);
+            assert_eq!(s.miss, 1);
+            assert_eq!(s.invalidate, 1);
+        }
+        let val: String = redis::cmd("GET")
+            .arg("key_1")
+            .query_async(&mut con)
+            .await
+            .unwrap();
+        assert_eq!(val, "1");
+        {
+            let s = con.get_cache_statistics();
+            assert_eq!(s.hit, 1);
+            assert_eq!(s.miss, 2);
+            assert_eq!(s.invalidate, 1);
+        }
+        Ok::<_, RedisError>(())
+    })
+    .unwrap();
+}
+
+#[test]
+fn test_cache_mget() {
+    use redis::ProtocolVersion;
+
+    let ctx = TestContext::new();
+    if ctx.protocol == ProtocolVersion::RESP2 {
+        return;
+    }
+
+    block_on_all(async move {
+        let mut con = ctx.multiplexed_async_connection_tokio_with_cache().await?;
+        let _: Vec<Option<String>> = redis::cmd("MGET")
+            .arg("key_1")
+            .arg("key_2")
+            .query_async(&mut con)
+            .await?;
+        {
+            let s = con.get_cache_statistics();
+            assert_eq!(s.hit, 0);
+            assert_eq!(s.miss, 2);
+        }
+
+        let _: Vec<Option<String>> = redis::cmd("MGET")
+            .arg("key_1")
+            .arg("key_2")
+            .query_async(&mut con)
+            .await?;
+        {
+            let s = con.get_cache_statistics();
+            assert_eq!(s.hit, 2);
+            assert_eq!(s.miss, 2);
+        }
+        let _: Option<String> = redis::cmd("GET").arg("key_1").query_async(&mut con).await?;
+        {
+            let s = con.get_cache_statistics();
+            assert_eq!(s.hit, 3);
+            assert_eq!(s.miss, 2);
+        }
+        Ok::<_, RedisError>(())
+    })
+    .unwrap();
+}
+
+#[test]
+fn test_cache_drops_and_guards() {
+    use redis::ProtocolVersion;
+
+    let ctx = TestContext::new();
+    if ctx.protocol == ProtocolVersion::RESP2 {
+        return;
+    }
+
+    block_on_all(async move {
+        let con = ctx
+            .multiplexed_async_connection_tokio_with_cache_config(CacheConfig::default())
+            .await?;
+
+        let mut v = vec![];
+        for i in 0..1000 {
+            let mut c = con.clone();
+            v.push(async move {
+                let x: RedisFuture<Option<u32>> = c.get("KEY");
+                tokio::time::timeout(Duration::from_micros(i), x).await
```

Comment:
why is the timeout needed? just for safety?

Reply:
Answered below.